core/src/main/java/com/linecorp/armeria/common/stream/EventLoopStreamMessage.java

```diff
@@ -0,0 +1,417 @@
+
+        if (state != State.OPEN) {
+            return;
+        }
+        state = State.CLOSED;
+
+        queue.add(closeEvent);
+        tryNotifyEvents();
+    }
+
+    @Override
+    public boolean isEmpty() {
+        return !isOpen() && !wroteAny;
+    }
+
+    @Override
+    public long demand() {
+        return demand;
+    }
+
+    @Override
+    public CompletableFuture<Void> whenComplete() {
+        return completionFuture;
+    }
+
+    @Override
+    public void subscribe(Subscriber<? super T> subscriber, EventExecutor executor,
+                          SubscriptionOption... options) {
+        requireNonNull(subscriber, "subscriber");
+        requireNonNull(executor, "executor");
+        requireNonNull(options, "options");
+        if (eventLoop.inEventLoop()) {
+            subscribe0(subscriber, executor, options);
+        } else {
+            eventLoop.execute(() -> subscribe0(subscriber, executor, options));
+        }
+    }
+
+    private void subscribe0(Subscriber<?> subscriber, EventExecutor executor,
+                            SubscriptionOption... options) {
+        assert eventLoop.inEventLoop();
+        if (subscription != null) {
+            notifySubscribeAndAbort(NoopSubscription.get(), subscriber,
+                                    abortedOrLate(subscription.subscriber()), executor);
+            return;
+        }
+        subscription = new SubscriptionImpl(this, subscriber, executor);
+        notifySubscribe0(subscription, subscriber, executor);
+        tryNotifyEvents();
+    }
+
+    @Override
+    public void abort() {
+        if (eventLoop.inEventLoop()) {
+            abort0(ABORTED_CLOSE);
+        } else {
+            eventLoop.execute(() -> abort0(ABORTED_CLOSE));
+        }
+    }
+
+    @Override
+    public void abort(Throwable cause) {
+        requireNonNull(cause, "cause");
+        if (eventLoop.inEventLoop()) {
+            abort0(new CloseEvent(cause));
+        } else {
+            eventLoop.execute(() -> abort0(new CloseEvent(cause)));
+        }
+    }
+
+    private void abort0(CloseEvent closeEvent) {
+        assert eventLoop.inEventLoop();
+        if (subscription == null) {
+            subscription = new SubscriptionImpl(this, AbortingSubscriber.get(closeEvent.cause()),
+                                                ImmediateEventExecutor.INSTANCE);
+        }
+        scheduleClose(closeEvent);
+    }
+
+    private void cancel() {
+        scheduleClose(CANCELLED_CLOSE);
+    }
+
+    private void scheduleClose(CloseEvent closeEvent) {
+        if (!eventLoop.inEventLoop()) {
+            eventLoop.execute(() -> scheduleClose(closeEvent));
+        }
+        assert subscription != null;
+        scheduleClose(subscription.subscriber(), closeEvent, subscription.executor());
+    }
+
+    private void scheduleClose(Subscriber<?> subscriber,
+                               CloseEvent closeEvent, EventExecutor executor) {
+        if (!eventLoop.inEventLoop()) {
```

Comment:
It would be nice to minimize and optimize to call `eventLoop.inEventLoop()` for private methods.